ecl/hql/hqlfold.cpp

```diff
@@ -1335,8 +1346,486 @@ IValue * foldExternalCall(IHqlExpression* expr, unsigned foldOptions, ITemplateC
+            IHqlExpression * value = child->queryChild(0);
+            if (value)
+            {
+                optionsStr.append("=");
+                value->queryValue()->getUTF8Value(optionsStr);
+            }
+        }
+    }
+    IHqlExpression *languageAttr = body->queryAttribute(languageAtom);
+    HqlExprArray noParams;
+    OwnedHqlExpr langLoadCall = createTranslatedExternalCall(NULL, languageAttr->queryChild(0), noParams);
+    IValue *plugin = foldExternalCall(langLoadCall, foldOptions, templateContext);
+    if (plugin == nullptr)
+        return NULL;
+
+    Owned<IEmbedContext> __plugin = (IEmbedContext *) plugin->getIntValue();  // We declared as int since ecl has no opinter type - not sure what the clean fix is here...
+    DummyContext dummyContext;
+    Owned<IEmbedFunctionContext> __ctx = __plugin->createFunctionContextEx(&dummyContext,flags,optionsStr.str());
+
+    if (!isImport)
+    {
+        IValue *query = body->queryChild(0)->queryValue();
+        assertex(query);
+        StringBuffer queryText;
+        query->getUTF8Value(queryText);
+        __ctx->compileEmbeddedScript(queryText.lengthUtf8(), queryText.str());
+    }
+
+    // process all the parameters passed in
+    unsigned numParam = expr->numChildren();
+    for(unsigned i = 0; i < numParam; i++)
+    {
+        IHqlExpression * curParam = expr->queryChild(i);            //NB: Already folded...
+        IHqlExpression * curArg = formals->queryChild(i);
+        assertex(curArg != NULL);
+        ITypeInfo * paramType = curArg->queryType();
+        assertex(paramType != NULL);
+        IValue * paramValue = curParam->queryValue();
+        assertex(curParam->isConstant());
+        unsigned paramSize = paramType->getSize();
+        IIdAtom * paramId = curArg->queryId();
+        const char * name = str(paramId);
+        // if (!foldOptions & HFOpreserveCaseExternalParameter)  // Does not seem to be reliably passed in - is there another way ?
+        //    name = str(lower(paramId));
+        switch (paramType->getTypeCode())
+        {
+        case type_int:
+        {
+            __int64 value = paramValue->getIntValue();
+            if (paramSize<8)
+            {
+                if (paramType->isSigned())
+                    __ctx->bindSignedSizeParam(name, paramSize, value);
+                else
+                    __ctx->bindUnsignedSizeParam(name, paramSize, value);
+            }
+            else
+            {
+                if (paramType->isSigned())
+                    __ctx->bindSignedParam(name, value);
+                else
+                    __ctx->bindUnsignedParam(name, value);
+            }
+            break;
+        }
+        case type_varstring:
+        {
+            StringBuffer value;
+            paramValue->getStringValue(value);
+            __ctx->bindVStringParam(name, value.str());
+            break;
+        }
+        case type_string:
+        {
+            StringBuffer value;
```

Comment:
Efficiency is not really a concern here. I'll have a go and see if it feels cleaner/clearer.